Silicon/Intel/IntelSiliconPkg/Feature/Flash/SpiFvbService/SpiFvbServiceMm.c

```diff
@@ -147,75 +155,134 @@ FvbInitialize (
     //
     // Make sure all FVB are valid and/or fix if possible
     //
-    for (Idx = 0;; Idx++) {
-      if (mPlatformFvBaseAddress[Idx].FvSize == 0 && mPlatformFvBaseAddress[Idx].FvBase == 0) {
+    for (Idx = 0; ; Idx++) {
+      if ((mPlatformFvBaseAddress[Idx].FvSize == 0) && (mPlatformFvBaseAddress[Idx].FvBase == 0)) {
         break;
       }
 
       BaseAddress = mPlatformFvBaseAddress[Idx].FvBase;
-      FvHeader = (EFI_FIRMWARE_VOLUME_HEADER *) (UINTN) BaseAddress;
+      FvHeader    = (EFI_FIRMWARE_VOLUME_HEADER *)(UINTN)BaseAddress;
 
       if (!IsFvHeaderValid (BaseAddress, FvHeader)) {
         BytesWritten = 0;
-        BytesErased = 0;
+        BytesErased  = 0;
         DEBUG ((DEBUG_ERROR, "ERROR - The FV in 0x%x is invalid!\n", FvHeader));
         FvHeader = NULL;
         Status   = GetFvbInfo (BaseAddress, &FvHeader);
         if (EFI_ERROR (Status)) {
           DEBUG ((DEBUG_WARN, "ERROR - Can't recovery FV header at 0x%x.  GetFvbInfo Status %r\n", BaseAddress, Status));
           continue;
         }
+
         DEBUG ((DEBUG_INFO, "Rewriting FV header at 0x%X with static data\n", BaseAddress));
         //
         // Spi erase
         //
-        BytesErased = (UINTN) FvHeader->BlockMap->Length;
-        Status = SpiFlashBlockErase( (UINTN) BaseAddress, &BytesErased);
+        BytesErased = (UINTN)FvHeader->BlockMap->Length;
+        Status      = SpiFlashBlockErase ((UINTN)BaseAddress, &BytesErased);
         if (EFI_ERROR (Status)) {
           DEBUG ((DEBUG_WARN, "ERROR - SpiFlashBlockErase Error  %r\n", Status));
           if (FvHeader != NULL) {
             FreePool (FvHeader);
           }
+
           continue;
         }
+
         if (BytesErased != FvHeader->BlockMap->Length) {
           DEBUG ((DEBUG_WARN, "ERROR - BytesErased != FvHeader->BlockMap->Length\n"));
           DEBUG ((DEBUG_INFO, " BytesErased = 0x%X\n Length = 0x%X\n", BytesErased, FvHeader->BlockMap->Length));
           if (FvHeader != NULL) {
             FreePool (FvHeader);
           }
+
           continue;
         }
-        BytesWritten = FvHeader->HeaderLength;
-        Status = SpiFlashWrite ((UINTN)BaseAddress, &BytesWritten, (UINT8*)FvHeader);
+
+        BytesWritten         = FvHeader->HeaderLength;
+        ExpectedBytesWritten = BytesWritten;
+        if (Idx != 0) {
+          Status = SpiFlashWrite ((UINTN)BaseAddress, &BytesWritten, (UINT8 *)FvHeader);
+        } else {

```

Comment:
Idx == 0 is VariableStore FV and here we should rewrite both FV header and VariableStore header.